orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -194,43 +198,135 @@ Status FromTensorProtos(
 
   constexpr size_t m_bytes = 1024 * 1024;
   size_t fbs_buffer_size = 0U;
+  size_t fbs_potential_external_buffer_size = 0;
   for (const auto& tensor_proto : trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      ORT_RETURN_IF_NOT(tensor_proto.external_data()[2].key() == "length",
+                        "Invalid external data for ", tensor_proto.name());
+      // external_data field is a dictionary of strings
+      // length is stored as a string
+      fbs_buffer_size += std::stoull(tensor_proto.external_data()[2].value());
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData) {  // assuming strings aren't trainable
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
+
   for (const auto& tensor_proto : non_trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      ORT_RETURN_IF_NOT(tensor_proto.external_data()[2].key() == "length",
+                        "Invalid external data for ", tensor_proto.name());
+      // external_data field is a dictionary of strings
+      // length is stored as a string
+      fbs_buffer_size += std::stoull(tensor_proto.external_data()[2].value());
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData &&
+        tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
 
+  // 2GB is the real limit, but fbs_buffer_size doesn't include space for initializer names and shapes.
+  // for now, arbitrarily use ~1.8GB as a conservative 'safe' value with a 10% buffer.
+  const bool use_external_data = fbs_buffer_size >= 1800 * m_bytes;
```

Comment:
nit: calculating this based on fbs_buffer_size before finalizing the value of fbs_buffer_size makes the code slightly harder to understand.

as use_external_data is a rough heuristic, determining it after aligning fbs_buffer_size shouldn't matter.